src/controller/CommissioneeDeviceProxy.h

```diff
@@ -0,0 +1,392 @@
+     */
+    void OnConnectionExpired(SessionHandle session) override;
+
+    /**
+     * @brief
+     *   This function is called when a message is received from the corresponding CHIP
+     *   device. The message ownership is transferred to the function, and it is expected
+     *   to release the message buffer before returning.
+     *
+     * @param[in] exchange      The exchange context the message was received
+     *                          on.  The Device guarantees that it will call
+     *                          Close() on exchange when it's done processing
+     *                          the message.
+     * @param[in] payloadHeader Reference to payload header in the message
+     * @param[in] msgBuf        The message buffer
+     */
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * exchange, const PayloadHeader & payloadHeader,
+                                 System::PacketBufferHandle && msgBuf) override;
+
+    /**
+     * @brief ExchangeDelegate implementation of OnResponseTimeout.
+     */
+    void OnResponseTimeout(Messaging::ExchangeContext * exchange) override;
+
+    /**
+     *  In case there exists an open session to the device, mark it as expired.
+     */
+    CHIP_ERROR CloseSession();
+
+    CHIP_ERROR Disconnect() override { return CloseSession(); }
+
+    /**
+     * @brief
+     *   Update address of the device.
+     *
+     *   This function will set new IP address and port of the device. Since the device settings might
+     *   have been moved from RAM to the persistent storage, the function will load the device settings
+     *   first, before making the changes.
+     *
+     * @param[in] addr   Address of the device to be set.
+     *
+     * @return CHIP_NO_ERROR if the address has been updated, an error code otherwise.
+     */
+    CHIP_ERROR UpdateAddress(const Transport::PeerAddress & addr);
+    /**
+     * @brief
+     *   Return whether the current device object is actively associated with a paired CHIP
+     *   device. An active object can be used to communicate with the corresponding device.
+     */
+    bool IsActive() const override { return mActive; }
+
+    void SetActive(bool active) { mActive = active; }
+
+    bool IsSecureConnected() const override { return IsActive() && mState == ConnectionState::SecureConnected; }
+
+    bool IsSessionSetupInProgress() const { return IsActive() && mState == ConnectionState::Connecting; }
+
+    void Reset();
+
+    NodeId GetDeviceId() const override { return mDeviceId; }
+
+    bool MatchesSession(SessionHandle session) const { return mSecureSession.HasValue() && mSecureSession.Value() == session; }
+
+    chip::Optional<SessionHandle> GetSecureSession() const override { return mSecureSession; }
+
+    Messaging::ExchangeManager * GetExchangeManager() const override { return mExchangeMgr; }
+
+    void SetAddress(const Inet::IPAddress & deviceAddr) { mDeviceAddress.SetIPAddress(deviceAddr); }
```

Comment:
Is this used anywhere? Can this also be a BLE? Why IPAddress and not PeerAddress? Ditto for the Get function - why not just return the PeerAddress directly?

Reply:
Looks like it's being used as a placeholder for DNS cache. For example, Python controller app calls `GetPeerAddressAndPort()` which in turn calls `GetAddress()`.

Also, for commissioning, `mDeviceAddress` stores the device's IP address (or stores type BLE).

So, somewhat convoluted, but it's needed for now. DNS cache will remove the first use-case. So we can probably remove `GetAddress()` at that point.